include/runtime_symbol_lookup.hpp

```diff
@@ -9,26 +9,133 @@
-                              SymbolTable &symbol_table);
+  struct hash_string {
+    // allow comparison between strings and string_views.
+    // This saves on some string creations
+    // as suggested by @sjanel
+    using is_transparent = void;
+    std::size_t operator()(const std::string &v) const {
+      return std::hash<std::string>{}(v);
+    }
+    std::size_t operator()(const char *v) const {
+      return std::hash<std::string_view>{}(v);
+    }
+
+    std::size_t operator()(const std::string_view &v) const {
+      return std::hash<std::string_view>{}(v);
+    }
+  };
+
+  using FailedCycle =
+      std::unordered_map<std::string, uint32_t, hash_string, std::equal_to<>>;
+  struct SymbolInfo {
+    SymbolMap _map;
+    FailedCycle _failed_cycle;
+  };
+  using PidUnorderedMap = std::unordered_map<pid_t, SymbolInfo>;
+
+  // Notes on JITDump strategy
+  //
+  // 1) Retrieve JITDump path
+  // Dso type will tell us that there is a JIT file.
+  // LLVM sources explain the logic about where we can find it. though we don't
+  // need that. The file is mmaped so we can get the path from there.
+  //
+  // We store in the DSOHdr the fact that we have a JITDump file for the pid.
+  //
+  // 2) Retrieve symbols
+  // Whenever we will come across the symbolisation of an unknown region,
+  // we use the runtime_symbol_lookup to check for existing symbols.
+  // If none are found, we parse the JITDump file if available.
+  // If not, we look for a perf-map file.
+  // Symbols are cached with the process's address.
+  //
+  DDRes fill_from_jitdump(std::string_view jitdump_path, pid_t pid,
+                          SymbolMap &symbol_map, SymbolTable &symbol_info);
+
+  DDRes fill_from_perfmap(int pid, SymbolMap &symbol_map,
+                          SymbolTable &symbol_table);
+
   FILE *perfmaps_open(int pid, const char *path_to_perfmap);
 
+  bool has_lookup_failure(const SymbolInfo &symbol_info,
+                          std::string_view path) const {
+    const auto it = symbol_info._failed_cycle.find(path);
+    if (it != symbol_info._failed_cycle.end()) {
+      // failure during this cycle
+      return it->second == _cycle_counter;
+    }
+    return false;
+  }
+
+  void flag_lookup_failure(SymbolInfo &symbol_info, std::string_view path) {
+    const auto it = symbol_info._failed_cycle.find(path);
+    // Written this way, we save up on creating strings
+    // only the slow path will create a string for the path
+    if (it != symbol_info._failed_cycle.end()) {
+      ++(it->second);
+    } else {
+      symbol_info._failed_cycle[std::string(path)] = _cycle_counter;
+    }
+    ++_stats._nb_failed_lookups;
+  }
+
+  bool should_skip_symbol(std::string_view symbol) const;
+
+  bool insert_or_replace(const std::string &symbol, ProcessAddress_t address,
+                         Offset_t size, SymbolMap &symbol_map,
+                         SymbolTable &symbol_table);
+
+  static constexpr std::array<const std::string_view, 1>
```

Comment:
Could be moved in `.cc` file

Reply:
I see value in having this more visible in header.